### DISCUSSION TITLE
Component Events (onAppear onDisappear)

### DISCUSSION THREAD

**Discussion by Aytharn** [fsprojects/Avalonia.FuncUI#384] [Q&A]
Dear Community, I am facing a lifetime issue with FuncUI, and the lifetime section in the docs doesn't really help.
My problem is that I need to use a barcode reader sending Reactive notification for different events through the COM port.
It does work but the problem is that without knowing when a specific component goes out of visibility and/or focus I cannot dispose the subscription or try to ignore the event...
Which means all my barcode enabled components keep reacting in the background while waiting to be garbage collected...

I really need something corresponding to onAppear and onDisappear methods of SwiftUI! :-(

**Comment by JaggerJo** [comment]
When a component is detached from the logical/visual tree effects get disposed. 

You can create an effect and return an `IDisposable`. In the `Dispose` method you can do the cleanup.

**Reply by Aytharn** [reply]
I think you are talking about what is In the section Hooks, and I suppose you mean consuming the attached type to a trigger... which is not really explained.

Problem is, I don't need an effect to watch http requests: I just need to know if the component is no more visible to the user, quite simple and basic (yet very important)

Thanks

**Reply by JaggerJo** [reply]
Is the component just not visible to the user anymore (component is located in a not selected tab) or is it not part of the UI anymore?

**Comment by Aytharn** [comment]
Just not visible, but could be both: behaviour is the same: even if not part of the visual tree: it is not disposed and the subscriptions are still being received, but let's assume it's just not visible and not focused: so? How can I detect that?

**Reply by JaggerJo** [reply]
https://github.com/AvaloniaUI/Avalonia/discussions/7876

**Reply by Aytharn** [reply]
Let me see if I understood it right: FuncUI absolutely needs to have onAppear and onDisappear implemented,  but there's no easy way to do it, at least with a method extension.

The proof is that in the carefully engineered piece of junk called UWP (one destined to the same bright future as Adobe Air, Flash or Silverlight) isn't even contemplated! 

Well, It explains a lot, excepted how to I make my UI respond to external events...

As I see it, ideally it should be done inside IComponentContext in a way you can do
`Component.create ("Login", fun ctx ->`
`        let state = ctx.useState _your_init_stuff`
`        ctx.onAppear(fun e -> ())`
`        ctx.onDisappear(fun e -> ())`

      
It's quite important in my needs, my app is massively using barcodes: yet connected on a comport so you don't have to click inside some textbox to read the scan as a text, which is a very common and sad way of proceeding see in many POS apps, one I really have no intention to implement. 
You see my point?

**Reply by JaggerJo** [reply]
I don't even understand your exact problem.

All I get is that you think FuncUI should have `onAppear/onDisappear`  hooks on components. Avalonia, which is the backing 
library for FuncUI has a way to check if a component is effectively visible - and there is [work going on](https://github.com/AvaloniaUI/Avalonia/pull/13972) so it's possible to listen
for changes. 

So my suggestion is: 
- Solve your issue in a different way - I'm sure there is one. 
- Contribute to Avalonia / FuncUI and implement the feature yourself.
